File: mobile/ios/RaverMVP/RaverMVP/Features/Profile/FollowListViewModel.swift
import Foundation

@MainActor
final class FollowListViewModel: ObservableObject {
    @Published var users: [UserSummary] = []
    @Published private(set) var phase: LoadPhase = .idle
    @Published var isLoading = false
    @Published var isRefreshing = false
    @Published var isLoadingMore = false
    @Published var bannerMessage: String?
    @Published var error: String?

    let userID: String
    let kind: FollowListKind

    private let repository: ProfileSocialRepository
    private var nextCursor: String?
    private var hasMore = true

    init(userID: String, kind: FollowListKind, repository: ProfileSocialRepository) {
        self.userID = userID
        self.kind = kind
        self.repository = repository
    }

    func load() async {
        if isLoading { return }
        isLoading = true
        let hadContent = !users.isEmpty
        if hadContent {
            isRefreshing = true
        } else {
            phase = .initialLoading
        }
        defer { isLoading = false }
        defer { isRefreshing = false }

        do {
            let page = try await fetchPage(cursor: nil)
            users = page.users
            nextCursor = page.nextCursor
            hasMore = page.nextCursor != nil
            phase = users.isEmpty ? .empty : .success
            bannerMessage = nil
            error = nil
        } catch {
            let message = error.userFacingMessage ?? L("列表加载失败，请稍后重试", "Failed to load the list. Please try again later.")
            if hadContent {
                bannerMessage = message
                phase = .success
            } else {
                phase = .failure(message: message)
            }
        }
    }

    func refresh() async {
        await load()
    }

    func loadMoreIfNeeded(currentUser: UserSummary) async {
        guard hasMore, !isLoadingMore, !isLoading else { return }
        guard currentUser.id == users.last?.id else { return }
        guard let cursor = nextCursor else { return }

        isLoadingMore = true
        defer { isLoadingMore = false }

        do {
            let page = try await fetchPage(cursor: cursor)
            let existing = Set(users.map(\.id))
            users.append(contentsOf: page.users.filter { !existing.contains($0.id) })
            nextCursor = page.nextCursor
            hasMore = page.nextCursor != nil
            bannerMessage = nil
            error = nil
        } catch {
            bannerMessage = error.userFacingMessage
        }
    }

    private func fetchPage(cursor: String?) async throws -> FollowListPage {
        switch kind {
        case .followers:
            return try await repository.fetchFollowers(userID: userID, cursor: cursor)
        case .following:
            return try await repository.fetchFollowing(userID: userID, cursor: cursor)
        case .friends:
            return try await repository.fetchFriends(userID: userID, cursor: cursor)
        }
    }

    func toggleFollow(user: UserSummary) async {
        do {
            let updated = try await repository.toggleFollow(userID: user.id, shouldFollow: !user.isFollowing)
            if let index = users.firstIndex(where: { $0.id == user.id }) {
                users[index] = updated
            }
        } catch {
            self.error = error.userFacingMessage
        }
    }
}

File: mobile/flutter_ravehub/packages/features/feature_profile/lib/src/follow_list/view_models/follow_list_view_model.dart
import 'package:flutter/foundation.dart';
import 'package:raver_design_system/raver_design_system.dart';
import 'package:raver_models/raver_models.dart';

import '../data/follow_api.dart';

class FollowListViewModel extends ChangeNotifier {
  FollowListViewModel({
    required FollowApi api,
    required this.userId,
    required this.listType,
  }) : _api = api;

  final FollowApi _api;
  final String userId;
  final String listType; // 'following', 'followers', or 'friends'

  LoadPhase<List<UserSummary>> _phase = const LoadPhase.loading();
  LoadPhase<List<UserSummary>> get phase => _phase;

  final List<UserSummary> _users = [];
  List<UserSummary> get users => List.unmodifiable(_users);

  int _currentPage = 1;
  int _totalPages = 1;
  bool _isLoadingMore = false;
  bool get isLoadingMore => _isLoadingMore;
  bool get canLoadMore => _currentPage < _totalPages;

  Future<void> load() async {
    _phase = const LoadPhase.loading();
    notifyListeners();

    try {
      final page = await _fetchPage(page: 1);

      _users
        ..clear()
        ..addAll(page.items);
      _currentPage = 1;
      _totalPages = page.pagination?.totalPages ?? 1;
      _phase =
          _users.isEmpty ? const LoadPhase.empty() : LoadPhase.success(_users);
    } catch (e) {
      _phase = LoadPhase.fromError(e);
    }
    notifyListeners();
  }

  Future<void> refresh() async {
    try {
      final page = await _fetchPage(page: 1);

      _users
        ..clear()
        ..addAll(page.items);
      _currentPage = 1;
      _totalPages = page.pagination?.totalPages ?? 1;
      _phase =
          _users.isEmpty ? const LoadPhase.empty() : LoadPhase.success(_users);
    } catch (e) {
      if (_users.isEmpty) _phase = LoadPhase.fromError(e);
    }
    notifyListeners();
  }

  Future<void> loadMore() async {
    if (_isLoadingMore || !canLoadMore) return;
    _isLoadingMore = true;
    notifyListeners();

    try {
      final nextPage = _currentPage + 1;
      final page = await _fetchPage(page: nextPage);

      _users.addAll(page.items);
      _currentPage = nextPage;
      _totalPages = page.pagination?.totalPages ?? _totalPages;
      _phase = LoadPhase.success(_users);
    } catch (_) {}
    _isLoadingMore = false;
    notifyListeners();
  }

  Future<BFFListPage<UserSummary>> _fetchPage({required int page}) {
    switch (listType) {
      case 'followers':
        return _api.fetchFollowers(userId: userId, page: page);
      case 'friends':
        return _api.fetchFriends(userId: userId, page: page);
      case 'following':
      default:
        return _api.fetchFollowing(userId: userId, page: page);
    }
  }
}
